driver-core/src/main/java/com/datastax/driver/core/AbstractEventDebouncer.java

```diff
@@ -0,0 +1,198 @@
+        this.executor = executor;
+        this.callback = callback;
+        this.delayMs = delayMs;
+        this.events = new ArrayBlockingQueue<T>(maxPendingEvents);
+    }
+
+    void start() {
+        logger.trace("Starting debouncer...");
+        started.countDown();
+        logger.trace("Debouncer started");
+        eventsLock.lock();
+        try {
+            if (!events.isEmpty()) {
+                logger.trace("Events were accumulated before the debouncer started: scheduling delivery now");
+                scheduleNewDelivery();
+            }
+        } finally {
+            eventsLock.unlock();
+        }
+    }
+
+    void stop() {
+        logger.trace("Stopping debouncer...");
+        stopped = true;
+        while (true) {
+            DeliveryAttempt previous = cancelPendingDelivery();
+            if (pendingDelivery.compareAndSet(previous, null)) {
+                break;
+            }
+        }
+        logger.trace("Debouncer stopped");
+    }
+
+    void eventReceived(T event) {
+        if(stopped) {
+            logger.trace("Debouncer is stopped, rejecting event: {}", event);
+            return;
+        }
+        checkNotNull(event);
+        logger.trace("Event received {}", event);
+        enqueueEvent(event);
+        scheduleNewDelivery();
+    }
+
+    private void enqueueEvent(T event) {
+        eventsLock.lock();
+        try {
+            while (!events.offer(event)) {
+                logger.trace("Event queue is full, blocking the caller and attempting to deliver immediately");
+                try {
+                    // if the queue is full and the debouncer is not ready,
+                    // we have no other choice than to block and wait
+                    started.await();
+                    // this can race with a scheduled delivery,
+                    // but in the worse case the number of delivered events will be zero
+                    // which is harmless
+                    deliverEvents();
+                } catch (InterruptedException e) {
+                    logger.warn("Interrupted while attempting to deliver an event");
+                    Thread.currentThread().interrupt();
+                }
+            }
+        } finally {
+            eventsLock.unlock();
+        }
+    }
+
+    private DeliveryAttempt cancelPendingDelivery() {
+        DeliveryAttempt previous = pendingDelivery.get();
+        if (previous != null)
+            previous.cancel();
+        return previous;
+    }
+
+    private void scheduleNewDelivery() {
+        while (!stopped) {
+            DeliveryAttempt previous = cancelPendingDelivery();
+            DeliveryAttempt next = new DeliveryAttempt();
+            if (pendingDelivery.compareAndSet(previous, next)) {
+                next.schedule();
+                break;
+            }
+        }
+    }
+
+    private void deliverEvents() throws InterruptedException {
```

Comment:
The normal case is to have this method called by a thread in the scheduled executor. In case of back pressure, the following threads are also likely to execute this method:
* IO thread (nio worker) in Cluster.handle(Response), if the queue is full (is it a problem? Note that the calling thread only executes the queue drain, once the events are drained they are processed in the main executor)
* Client thread or cluster-reconnection pool in ControlConnection.tryConnect(Host, boolean) (only happens at initial connection attempt)